lib/provider.js

```diff
@@ -218,36 +219,48 @@ class StorageProvider {
   //   callback <Function>
   //     error <Error> | <null>
   //     result <any>
-  execute(category, action, actionArgs, callback) {
-    let foundAction;
-    if (category === null) {
-      foundAction = this.schema.actions.get(action);
-    } else {
-      const categorySchema = this.schema.get(category);
-      if (!categorySchema) {
+  //   permissionChecker - <Function>, optional
+  //     category - <string>
+  //     action - <string>
+  //     callback - <Function>
+  //       err - <Error> | <null>
+  execute(category, action, actionArgs, callback, permissionChecker) {
+    runIfFn(permissionChecker, category, action, err => {
+      if (err) {
+        callback(err);
+        return;
+      }
+
+      let foundAction;
+      if (category === null) {
+        foundAction = this.schema.actions.get(action);
+      } else {
+        const categorySchema = this.schema.categories.get(category);
+        if (!categorySchema) {
+          callback(new GSError(errorCodes.NOT_FOUND));
+          return;
+        }
+        foundAction = categorySchema.actions.get(action);
+      }
+      if (!foundAction) {
         callback(new GSError(errorCodes.NOT_FOUND));
         return;
       }
-      foundAction = categorySchema.actions.get(action);
-    }
-    if (!foundAction) {
-      callback(new GSError(errorCodes.NOT_FOUND));
-      return;
-    }
-    const args = actionArgs[1];
-    const error = this.schema.validate('action', foundAction, args);
-    if (error) {
-      process.nextTick(
-        callback,
-        new GSError(
-          errorCodes.INVALID_SCHEMA,
-          `Invalid arguments provided: ${error}`
-        )
-      );
-      return;
-    }
-    const { Execute: execute } = foundAction.definition;
-    execute(...actionArgs, callback);
+      const args = actionArgs[1];
```

Comment:
Should we perhaps use `actionArgs[actionArgs.length - 1]`? (And therefore state that args must be last element of the array and allow us to add other parameters without changing this function)

Reply:
We can change it in a separate PR, I haven't changed the code here (except for a

Reply:
We can change it in a separate PR, here I have only reordered the existing code without changing the behavior.